
.. _10X:

Parsing 10X VDJ output
================================================================================

Example Data
--------------------------------------------------------------------------------

Example data can be found by downloading Ig VDJ data provided by 
`10xgenomics <https://support.10xgenomics.com/single-cell-vdj/datasets/3.0.0/vdj_v1_hs_pbmc2_b>`__
which correspond to files generated by cellranger 2.0 or cellranger 3.0 pipelines. 

Merging with 10X annotations
--------------------------------------------------------------------------------

Changeo-O does not currently support a formal :ref:`MakeDb` option for 10X VDJ data.  

.. note::

    To generate, formatted data from 10X VDJ FASTA files, please see the 
    `Using IgBLAST documentation <https://changeo.readthedocs.io/en/version-0.4.5/examples/igblast.html>`__ 
    for instructions on how to build the database in a more general case. Shown below is an example of how
    to perform these steps. You must have all associated commands in your ``PATH`` and the appropriate 
    directories created::

	AssignGenes.py igblast -s 10x.fasta -b ~/share/igblast \
	   --organism human --loci ig --format blast
	MakeDb.py igblast -i 10x.fmt7 -s 10x.fasta \
	   -r IMGT_Human_IGHV.fasta IMGT_Human_IGHD.fasta IMGT_Human_IGHJ.fasta \
	   --regions --scores

    Once databases are built, they can then be merged with 10X VDJ CSV annotations. 


An in-line python command can be used as a temporary workaround for merging 10X annotations with the output of :ref:`MakeDb`. See ` merge10x.py <https://bitbucket.org/kleinstein/immcantation/src/tip/scripts/merge10x.py>`__ for example use ::

	merge10x.py 10x_db-pass.tab 10x_annotations.csv 10x_annotated_db-pass.tab

.. warning::

    The resulting table contains two V and J segment references corresponding to those
    generated by IgBLAST and those generated by cellranger vdj.