
.. _10X:

Parsing 10X Genomics V(D)J data
================================================================================

Example data
--------------------------------------------------------------------------------

10X Genomics provides an example data set of Ig V(D)J processed by the Cell
Ranger pipeline, which is available for download from their
`Single Cell Immune Profiling <https://support.10xgenomics.com/single-cell-vdj/datasets/3.0.0/vdj_v1_hs_pbmc2_b>`__
support site.

Converting 10X VDJ data into Change-O format
--------------------------------------------------------------------------------

To process 10X VDJ data, a combination of :ref:`AssignGenes` and :ref:`MakeDb` can be used to 
generate a TSV file conforming to :ref:`Change-O <Standard>` that incorporates 
annotation information provided by the Cell Ranger pipeline. The :option:`--af filtered_contig_annotations.csv <MakeDb>` 
specifies the path of the contig annotations file generated by ```cellranger vdj``` (found in the ```outs``` directory). 
The :option:`--ai contig_id <MakeDb>` specifies the column to use for matching sequence IDs between filtered_contig.fasta and filtered_contig_annotations.csv. 

Generate Change-O formatted data from the 10X V(D)J FASTA files using the
steps below::

	AssignGenes.py igblast -s filtered_contig.fasta -b ~/share/igblast \
	   --organism human --loci ig --format blast
	MakeDb.py igblast -i filtered_contig_igblast.fmt7 -s filtered_contig.fasta \
	   -r IMGT_Human_IGHV.fasta IMGT_Human_IGHD.fasta IMGT_Human_IGHJ.fasta \
	   --af filtered_contig_annotations.csv --ai contig_id \
	   --regions --scores

all_contig.fasta can be exchanged for filtered_contig.fasta, and all_contig_annotations.csv can be exchanged 
for filtered_contig_annotations.csv.

.. seealso::

    See the :ref:`IgBLAST usage guide <IgBLAST>` for further details regarding
    the setup and use of IgBLAST.

.. warning::

    The resulting table contains two V and J segment assignments corresponding to
    those generated by IgBLAST and those generated by Cell Ranger. 
    
Combining 10X VDJ data with Change-O format
--------------------------------------------------------------------------------

`Change-O <Standard>` compliant TSV files can also be merged without the use of :ref:`MakeDb`.
`merge10x.py <https://bitbucket.org/kleinstein/immcantation/src/tip/scripts/merge10x.py>`__,
can be used to perform the merge is provided on the
`Immcantation Bitbucket repository <https://bitbucket.org/kleinstein/immcantation>`__
in the ``scripts`` directory. The following example shows how to merge the
Change-O file (``10x_igblast_db-pass.tab``) with a 10X annotation file
(``filtered_contig_annotations.csv``) and save it as ``10x_annotated_db-pass.tab``::

	merge10x.py 10x_igblast_db-pass.tab filtered_contig_annotations.csv 10x_annotated_db-pass.tab

.. warning::
    Using `merge10x.py <https://bitbucket.org/kleinstein/immcantation/src/tip/scripts/merge10x.py>`__, 
    does not guarantee the generation of :ref:`Change-O <Standard>` or :ref:`AIRR <Standard>` compliant TSV files. 
